### PR TITLE
Resource loss prob

### PR DESCRIPTION
**Pull request by gregsoftmax**
Added "resource_loss_p" variable in game config that is a probability that every individual resource held by an agent disappears from their inventory every timestep

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1211068593737466)

### CODE REVIEW

File: mettagrid/src/metta/mettagrid/mettagrid_c.cpp [hunk 1]
```diff
@@ -522,6 +523,39 @@ void MettaGrid::_step(py::array_t<ActionType, py::array::c_style> actions) {
...
+   for (auto& agent : _agents) {
+     if (_resource_loss_prob > 0.0f) {
+       // For every resource in an agent's inventory, it should disappear with probability _resource_loss_prob
+       std::vector<InventoryItem> items_to_remove;
+       for (auto& [item, qty] : agent->inventory) {
+         if (qty > 0) {
+           int lost = 0;
+           for (int i = 0; i < qty; ++i) {
+             // std::generate_canonical<float, 10>(_rng) produces a random float in [0, 1)
+             if (std::generate_canonical<float, 10>(_rng) < _resource_loss_prob) {
```
Comment (daveey) [inline mettagrid/src/metta/mettagrid/mettagrid_c.cpp:536]: lost = int(rnd()*count)

So that you don't need a loop

E[num_lost] = n × p_loss

File: mettagrid/src/metta/mettagrid/mettagrid_c.cpp [hunk 2]
```diff
@@ -522,6 +523,39 @@ void MettaGrid::_step(py::array_t<ActionType, py::array::c_style> actions) {
...
+             if (actual_lost > 0) {
+               qty -= actual_lost;
+               if (qty <= 0) {
+                 items_to_remove.push_back(item);
+               }
+             }
+           }
+         }
+       }
+       // Remove items that reached zero quantity
```
Comment (daveey) [inline mettagrid/src/metta/mettagrid/mettagrid_c.cpp:552]: Use update_inventory

**Review by daveey — CHANGES_REQUESTED** [review]

File: mettagrid/src/metta/mettagrid/mettagrid_c.cpp [hunk 3]
```diff
@@ -522,6 +523,29 @@ void MettaGrid::_step(py::array_t<ActionType, py::array::c_style> actions) {
...
+       // Make a real copy of the agent's inventory map to avoid iterator invalidation
+       const auto inventory_copy = agent->inventory;
+       for (const auto& [item, qty] : inventory_copy) {
+         if (qty > 0) {
+           double loss = _resource_loss_prob * qty;
+           int lost = static_cast<int>(std::floor(loss));
+           // With probability equal to the fractional part, lose one more
+           if (std::generate_canonical<float, 10>(_rng) < loss - lost) {
+             lost += 1;
+           }
```
[the remark this answers is not in the document]
Comment (daveey) [reply mettagrid/src/metta/mettagrid/mettagrid_c.cpp:539]: That's even better
